### PR TITLE
sql,cli: add SHOW FUNCTIONS command and \df metacommand

### LINKED ISSUE #88597: cli: `\df` command to list or show all user defined functions
It would be nice to have a dedicated command to show all user defined functions in a database or schema.
- PostgreSQL's psql has the [`\df` command](https://www.postgresql.org/docs/current/app-psql.html#APP-PSQL-META-COMMANDS).
  
  ```
  michae2=# \df
                         List of functions
   Schema | Name | Result data type | Argument data types  | Type
  --------+------+------------------+----------------------+------
   public | f2   | integer          |                      | func
   public | g    | integer          | a integer, b integer | func
   public | g    | integer          | i integer            | func
   public | m    | character        | f double precision   | func
   public | m    | character        | i integer            | func
   public | p    | boolean          | i integer            | func
  (6 rows)
  ```

- MySQL has the [`SHOW FUNCTION STATUS` statement](https://dev.mysql.com/doc/refman/8.0/en/show-function-status.html).

We do include user-defined functions in the output of `\hf` but this only gives the function name, not any additional information:

```
demo@127.0.0.1:26257/defaultdb> CREATE FUNCTION z () RETURNS INT IMMUTABLE LANGUAGE SQL AS $$ SELECT 1 $$;
CREATE FUNCTION


Time: 24ms total (execution 24ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> \hf
                            function
----------------------------------------------------------------
  _st_contains
  _st_containsproperly
  _st_coveredby
...
  with_max_staleness
  with_min_timestamp
  xor_agg
  z
(919 rows)


Time: 20ms total (execution 17ms / network 3ms)

demo@127.0.0.1:26257/defaultdb> \hf z
no help available for "z".
Try \hf with no argument to see available help.
```

As a workaround, we can query `pg_catalog.pg_proc`:

```
demo@127.0.0.1:26257/defaultdb> SELECT proname, prosrc FROM pg_catalog.pg_namespace n JOIN pg_catalog.pg_proc ON pronamespace = n.oid WHERE nspname = 'public';
  proname |  prosrc
----------+------------
  z       | SELECT 1;
(1 row)


Time: 17ms total (execution 17ms / network 0ms)
```

But this is a little annoying, and it would be nice to match psql.

This is on `v22.2.0-alpha.3-550-g099da40b73` (current tip of master).

Jira issue: CRDB-19851

Epic CRDB-17785

### PR DESCRIPTION
**Pull request by rafiss**
fixes #88597

Release note (sql change): Added the SHOW FUNCTIONS command, which lists
user-defined functions. The SHOW FUNCTIONS FROM <schema> syntax is
supported too.

Release note (cli change): The \df metacommand was added to the SQL
shell, which will list all user-defined functions in the current
database.

### CODE REVIEW

**Comment by knz** [comment]
Do you want to do `\df` this way or rather add to #88061?

**Comment by rafiss** [comment]
I'd like to backport this to v22.2, so I vote for this way and we can use the new approach in a future PR

File: pkg/sql/delegate/show_functions.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+// Privileges: None.
+//
+//	Notes: postgres does not have a SHOW FUNCTIONS statement.
+func (d *delegator) delegateShowFunctions(n *tree.ShowFunctions) (tree.Statement, error) {
+	flags := cat.Flags{AvoidDescriptorCaches: true}
+	_, name, err := d.catalog.ResolveSchema(d.ctx, flags, &n.ObjectNamePrefix)
+	if err != nil {
+		return nil, err
+	}
+	// If we're resolved a one-part name into <db>.public (which is the behavior
```
Comment (chengxiong-ruan) [inline pkg/sql/delegate/show_functions.go:33]: FWIW, I saw it says `Compatibility with CockroachDB v1.1: use D.public.T instead.` But I agree with the override you're doing here.

File: pkg/sql/delegate/show_functions.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	// name to have an explicit catalog but no explicit schema. This would arise
+	// when doing SHOW FUNCTIONS FROM <db>. Without this logic, we would not show the
+	// tables from other schemas than public.
+	if name.ExplicitSchema && name.ExplicitCatalog && name.SchemaName == tree.PublicSchemaName &&
+		n.ExplicitSchema && !n.ExplicitCatalog && n.SchemaName == name.CatalogName {
+		name.SchemaName, name.ExplicitSchema = "", false
+	}
+	var schemaClause string
+	if name.ExplicitSchema {
+		schema := lexbase.EscapeSQLString(name.Schema())
```
Comment (chengxiong-ruan) [inline pkg/sql/delegate/show_functions.go:44]: hmm, could you explain why do we need to escape the resolved names here?

**Review by chengxiong-ruan — APPROVED** [review]
Nice work! just one nit question.

File: pkg/sql/delegate/show_functions.go (hunk 2 above)
Comment (rafiss) [reply pkg/sql/delegate/show_functions.go:44]: `escapeSQLString` also takes care of adding single-quotes around the `go` string, so therefore it becomes safe to use in the `AND n.nspame` clause below.

File: pkg/sql/delegate/show_functions.go (hunk 1 above)
Comment (rafiss) [reply pkg/sql/delegate/show_functions.go:33]: ah, i had copied this comment over from `show_tables.go`, but yes that sounds right

**Comment by rafiss** [comment]
tftr!

bors r=chengxiong-ruan

**Pull request merged**